runtime/altair/src/lib.rs

```diff
@@ -1501,6 +1604,101 @@ pub type Executive = frame_executive::Executive<
 	UpgradeAltair1027,
 >;
 
+mod session_key_migration {
+	use frame_support::{pallet_prelude::*, storage_alias};
+
+	use super::*;
+	pub struct ExtendSessionKeys;
+
+	impl_opaque_keys! {
+		pub struct OldSessionKeys {
+			pub aura: Aura,
+		}
+	}
+
+	#[storage_alias]
+	type QueuedKeys<T: pallet_session::Config> = StorageValue<
+		pallet_session::Pallet<T>,
+		// hex_literal::hex!("cec5070d609dd3497f72bde07fc96ba0"),
+		Vec<(AccountId, OldSessionKeys)>,
+		ValueQuery,
+	>;
+
+	impl frame_support::traits::OnRuntimeUpgrade for ExtendSessionKeys {
+		#[cfg(feature = "try-runtime")]
+		fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+			let queued: Vec<(AccountId, OldSessionKeys)> = QueuedKeys::<Runtime>::get();
+
+			assert!(
+				!queued.len().is_zero(),
+				"Queued session keys must never be empty"
+			);
+
+			Ok(queued.encode())
+		}
+
+		fn on_runtime_upgrade() -> frame_support::weights::Weight {
+			use sp_runtime::SaturatedConversion;
+
+			// We must not bump the storage version of pallet_session to stay in sync with
+			// Substrate. As this migration is tightly coupled to the initialization of
+			// BlockRewards, it should be sane to trigger the session migration based on the
+			// BlockRewards storage version.
+			if BlockRewards::on_chain_storage_version() == StorageVersion::new(0) {
+				Session::upgrade_keys::<OldSessionKeys, _>(|_, old_key| SessionKeys {
+					aura: old_key.aura.clone(),
+					block_rewards: old_key.aura,
+				});
+
+				let n: u64 = pallet_session::Validators::<Runtime>::get()
+					.len()
+					.saturated_into();
+
+				<Runtime as frame_system::Config>::DbWeight::get()
+					.reads_writes(n.saturating_add(1), n)
+			} else {
+				<Runtime as frame_system::Config>::DbWeight::get().reads(1)
+			}
+		}
+
+		#[cfg(feature = "try-runtime")]
+		fn post_upgrade(pre_state: Vec<u8>) -> Result<(), &'static str> {
+			let pre_state: Vec<(AccountId, OldSessionKeys)> =
+				Decode::decode(&mut pre_state.as_slice())
+					.expect("Pre state parameter should be generated by pre-upgrade hooks");
+
+			let queued: Vec<(AccountId, SessionKeys)> =
+				pallet_session::QueuedKeys::<Runtime>::get();
+
+			assert!(
+				pre_state.len() == queued.len(),
+				"Mismatch in number of session keys: pre {} vs post {}",
+				pre_state.len(),
+				queued.len()
+			);
+
+			for i in 0..queued.len() {
+				assert!(
+					&queued[i].0 == &pre_state[i].0,
+					"Failed to match old validator key at index {}: old {:?} vs {:?} new",
+					i,
+					&pre_state[i].0,
+					&queued[i].0
+				);
+				assert!(
+					&queued[i].1.aura == &pre_state[i].1.aura,
+					"Failed to match old aura key at index {}. old {:?} vs {:?} new",
+					i,
+					&pre_state[i].1.aura,
+					&queued[i].1.aura,
+				);
+			}
```

Comment:
Can we do a `forEach` of sorts instead of `for i n 0.. x.len()`? I find that easier to read and safer as well. Since we want to compare the `queued` with the  `pre_state`, we can just `zip` the two, forEach on the zipped tuples, and go with that :)